### PR TITLE
Rework and add unit tests

### PR DESCRIPTION
**Pull request by dehall**
Adds unit testing for most of the core GMF engine, ie States and Logic. These unit tests were largely copied and translated from the ruby version.

In support of these unit tests, a lot had to be refactored and reworked. 
This includes rewriting States to be OO instead of using a single state class and an enum switch. Notably, the new State subclasses initialize themselves from the JSON definition, but then do not store the JsonObject to reference at runtime. Any fields they care about must be specifically defined and stored at initialization.

Also adds a "DeathModule" which runs after the rest of the modules, to add a death certificate with cause of death.

### CODE REVIEW

**Comment by dehall** [comment]
@jawalonoski should we track test code coverage metrics? I can add jacoco as well pretty easily.  we can also use https://codecov.io/  but it needs a synthetichealth admin to setup the account

File: src/main/java/org/mitre/synthea/engine/Module.java [hunk 1]
```diff
@@ -119,30 +122,24 @@ public static Module getModuleByPath(String path) {
 	
 	public String name;
 	public boolean submodule;
-	public List<String> remarks;
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/engine/Module.java]: Why did you get rid of the remarks? Once we have an editor up and running, I'm sure these will become increasingly useful.

File: src/main/java/org/mitre/synthea/world/agents/Person.java [hunk 2]
```diff
@@ -71,25 +77,28 @@ public double rand() {
...
+
+	public Period age(long time)
+	{
+		Period age = Period.ZERO;
+		
+		if(attributes.containsKey(BIRTHDATE)) 
+		{
+			LocalDate now = Instant.ofEpochMilli(time).atZone(ZoneId.systemDefault()).toLocalDate();
+			LocalDate birthdate = Instant.ofEpochMilli((long)attributes.get(BIRTHDATE)).atZone(ZoneId.systemDefault()).toLocalDate();
+			age = Period.between(birthdate, now);
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/world/agents/Person.java]: Sigh... There is something nice about subtraction. Simplicity is lost now in favor of ... Java "robustness"? Oh well.

File: src/main/java/org/mitre/synthea/engine/Module.java (hunk 1 above)
Comment (dehall) [reply src/main/java/org/mitre/synthea/engine/Module.java]: I was conflicted about remarks. On one hand I don't want the remarks to be actually used for anything, other than pure documentation. But you're right they will be useful in the editor. 
The only thing is that we can have remarks anywhere, not just at the state level. I'll take a closer look shortly.

File: src/main/java/org/mitre/synthea/world/agents/Person.java (hunk 2 above)
Comment (dehall) [reply src/main/java/org/mitre/synthea/world/agents/Person.java]: Simplicity vs accuracy. There are too many edge cases or way to get things wrong if we just deal with pure milliseconds.

Comment (jawalonoski) [reply src/main/java/org/mitre/synthea/world/agents/Person.java]: I'm not asking you to change it. But, I disagree. Time is based on math. There is nothing to get wrong. Timezones and calendars are irrelevant -- we know indisputable facts about time that are just basic mathematics. 1000ms = 1s = 1/60 min, etc.

**Pull request merged**